File: 01_GOF23/DP_examples/Go/02_Structural/Flyweight/flyweight_factory.go
package Flyweight

import "fmt"

// 享元工厂
type FlyweightFactory struct {
	shapeDic      map[string]map[float64]*CircleData
	circleFactory CircleFactory
}

// 单例实例
var instance *FlyweightFactory

// 获取单例实例
func GetFlyweightFactory() *FlyweightFactory {
	if instance == nil {
		instance = &FlyweightFactory{
			shapeDic:      make(map[string]map[float64]*CircleData),
			circleFactory: CircleFactory{},
		}
	}
	return instance
}

// 获取圆形
func (ff *FlyweightFactory) GetCircle(color string, radius, x, y float64) *Circle {
	var circle *Circle

	if dic, ok := ff.shapeDic[color]; ok {
		if data, ok := dic[radius]; ok {
			return ff.circleFactory.GetCircleByData(data, x, y)
		} else {
			circle = ff.circleFactory.GetCircle(color, radius, x, y)
			dic[radius] = circle.CircleData
		}
	} else {
		dic := make(map[float64]*CircleData)
		circle = ff.circleFactory.GetCircle(color, radius, x, y)
		dic[radius] = circle.CircleData
		ff.shapeDic[color] = dic
	}
	fmt.Println("Create a New Circle Filled with " + color)
	return circle
}

// 获取特殊圆形
func (ff *FlyweightFactory) GetSpecialCircle(color string, radius, x, y float64) *CircleAddition {
	return ff.circleFactory.GetSpacialCircle(ff.GetCircle(color, radius, x, y))
}

File: 01_GOF23/DP_examples/Python/02_Structural/Flyweight/FlyweightFactory.py
from Flyweight import Circle
from UnsharedFlyweight import CircleAddition, CircleFactory

# 享元工厂
class FlyweightFactory:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(FlyweightFactory, cls).__new__(cls)
            cls._instance.shape_dic = {}
            cls._instance.circle_factory = CircleFactory()
        return cls._instance
    
    def get_circle(self, color: str, radius: float, x: float, y: float) -> Circle:
        circle = None
        
        if color in self.shape_dic:
            if radius in self.shape_dic[color]:
                data = self.shape_dic[color][radius]
                return self.circle_factory.get_circle_by_data(data, x, y)
            else:
                circle = self.circle_factory.get_circle(color, radius, x, y)
                self.shape_dic[color][radius] = circle.data
        else:
            self.shape_dic[color] = {}
            circle = self.circle_factory.get_circle(color, radius, x, y)
            self.shape_dic[color][radius] = circle.data
        
        print(f"Create a New Circle Filled with {color}")
        return circle
    
    def get_special_circle(self, color: str, radius: float, x: float, y: float) -> CircleAddition:
        circle = self.get_circle(color, radius, x, y)
        return self.circle_factory.get_spacial_circle(circle)
